File: src/interpolatepy/quaternion/_shooting_solver.py
"""Multiple shooting for natural Riemannian cubics on unit quaternions.

Each segment uses local time in [0, 1] and nine unknowns (v, a, c).
The pure-quaternion Lie bracket is [v, a] = 2 cross(v, a).
"""

from __future__ import annotations

import numpy as np
from scipy.sparse import csc_matrix
from scipy.sparse import lil_matrix
from scipy.sparse.linalg import spsolve


_SMALL_ANGLE = 1e-7


def skew(values: np.ndarray) -> np.ndarray:
    """Return a batch of cross-product matrices."""
    result = np.zeros((*values.shape[:-1], 3, 3))
    result[..., 0, 1] = -values[..., 2]
    result[..., 0, 2] = values[..., 1]
    result[..., 1, 0] = values[..., 2]
    result[..., 1, 2] = -values[..., 0]
    result[..., 2, 0] = -values[..., 1]
    result[..., 2, 1] = values[..., 0]
    return result


def left_matrix(quaternions: np.ndarray) -> np.ndarray:
    """Matrices for left quaternion multiplication, in scalar-first order."""
    result = np.zeros((len(quaternions), 4, 4))
    result[:, 0, 0] = quaternions[:, 0]
    result[:, 0, 1:] = -quaternions[:, 1:]
    result[:, 1:, 0] = quaternions[:, 1:]
    result[:, 1:, 1:] = quaternions[:, :1, None] * np.eye(3) + skew(quaternions[:, 1:])
    return result


def rotation_residual(values: np.ndarray) -> tuple[np.ndarray, np.ndarray]:
    """Quaternion logarithms and their analytic derivatives, modulo sign."""
    signs = np.where(values[:, 0] < 0.0, -1.0, 1.0)
    canonical = values * signs[:, None]
    scalar, vector = canonical[:, 0], canonical[:, 1:]
    radius = np.linalg.norm(vector, axis=1)
    squared_norm = scalar**2 + radius**2
    small = radius < _SMALL_ANGLE
    factor = np.empty(len(values))
    coefficient = np.empty(len(values))
    factor[small] = 1.0 / scalar[small] - radius[small] ** 2 / (3.0 * scalar[small] ** 3)
    coefficient[small] = -2.0 / (3.0 * scalar[small] ** 3)
    factor[~small] = np.arctan2(radius[~small], scalar[~small]) / radius[~small]
    coefficient[~small] = (scalar[~small] / squared_norm[~small] - factor[~small]) / radius[~small] ** 2
    derivative = np.empty((len(values), 3, 4))
    derivative[:, :, 0] = -vector / squared_norm[:, None]
    derivative[:, :, 1:] = (
        factor[:, None, None] * np.eye(3) + coefficient[:, None, None] * vector[:, :, None] * vector[:, None, :]
    )
    return factor[:, None] * vector, derivative * signs[:, None, None]


def _rhs(
    state: np.ndarray, constants: np.ndarray, sensitivity: np.ndarray | None
) -> tuple[np.ndarray, np.ndarray | None]:
    """Evaluate the cubic ODE and its variational equations."""
    quaternion, velocity, acceleration = state[:, :4], state[:, 4:7], state[:, 7:10]
    result = np.empty_like(state)
    result[:, 0] = -np.sum(quaternion[:, 1:] * velocity, axis=1)
    result[:, 1:4] = quaternion[:, :1] * velocity + np.cross(quaternion[:, 1:], velocity)
    result[:, 4:7] = acceleration
    result[:, 7:10] = constants - 2.0 * np.cross(velocity, acceleration)
    if sensitivity is None:
        return result, None

    right = np.zeros((len(state), 4, 4))
    right[:, 0, 1:] = -velocity
    right[:, 1:, 0] = velocity
    right[:, 1:, 1:] = -skew(velocity)
    product = left_matrix(quaternion)[:, :, 1:]
    derivative = np.empty_like(sensitivity)
    derivative[:, :4] = right @ sensitivity[:, :4] + product @ sensitivity[:, 4:7]
    derivative[:, 4:7] = sensitivity[:, 7:10]
    derivative[:, 7:10] = 2.0 * skew(acceleration) @ sensitivity[:, 4:7] - 2.0 * skew(velocity) @ sensitivity[:, 7:10]
    derivative[:, 7:10, 6:9] += np.eye(3)
    return result, derivative


def _advance_sensitivity(
    sensitivity: np.ndarray | None, derivative: np.ndarray | None, step: float
) -> np.ndarray | None:
    if sensitivity is None:
        return None
    assert derivative is not None
    return sensitivity + step * derivative


def rk4_step(
    state: np.ndarray,
    constants: np.ndarray,
    step: float,
    sensitivity: np.ndarray | None = None,
) -> tuple[np.ndarray, np.ndarray | None, np.ndarray]:
    """Advance states and optional sensitivities; accumulate squared acceleration."""
    k1, d1 = _rhs(state, constants, sensitivity)
    y2 = state + 0.5 * step * k1
    k2, d2 = _rhs(y2, constants, _advance_sensitivity(sensitivity, d1, 0.5 * step))
    y3 = state + 0.5 * step * k2
    k3, d3 = _rhs(y3, constants, _advance_sensitivity(sensitivity, d2, 0.5 * step))
    y4 = state + step * k3
    k4, d4 = _rhs(y4, constants, _advance_sensitivity(sensitivity, d3, step))
    result = state + (step / 6.0) * (k1 + 2.0 * k2 + 2.0 * k3 + k4)
    norm = np.linalg.norm(result[:, :4], axis=1)
    result[:, :4] /= norm[:, None]
    if sensitivity is not None:
        assert d1 is not None
        assert d2 is not None
        assert d3 is not None
        assert d4 is not None
        updated = sensitivity + (step / 6.0) * (d1 + 2.0 * d2 + 2.0 * d3 + d4)
        quaternion = result[:, :4]
        radial = np.einsum("mi,mij->mj", quaternion, updated[:, :4])
        updated[:, :4] = (updated[:, :4] - quaternion[:, :, None] * radial[:, None, :]) / norm[:, None, None]
        sensitivity = updated
    energy = (step / 6.0) * (
        np.sum(state[:, 7:10] ** 2, axis=1)
        + 2.0 * np.sum(y2[:, 7:10] ** 2, axis=1)
        + 2.0 * np.sum(y3[:, 7:10] ** 2, axis=1)
        + np.sum(y4[:, 7:10] ** 2, axis=1)
    )
    return result, sensitivity, energy


def integrate(
    parameters: np.ndarray,
    starts: np.ndarray,
    steps: int,
    *,
    jacobian: bool = False,
    store_nodes: bool = False,
) -> tuple[np.ndarray, np.ndarray | None, np.ndarray, np.ndarray]:
    """Integrate every shooting segment together on its local unit interval."""
    state = np.column_stack((starts, parameters[:, :6]))
    sensitivity = None
    if jacobian:
        sensitivity = np.zeros((len(starts), 10, 9))
        sensitivity[:, 4:10, :6] = np.eye(6)
    nodes = [state.copy()] if store_nodes else []
    energy = np.zeros(len(starts))
    with np.errstate(over="ignore", invalid="ignore", divide="ignore"):
        for _ in range(steps):
            state, sensitivity, increment = rk4_step(state, parameters[:, 6:9], 1.0 / steps, sensitivity)
            energy += increment
            if not np.all(np.isfinite(state)):
                raise RuntimeError("Multiple shooting integration diverged")
            if store_nodes:
                nodes.append(state.copy())
    return state, sensitivity, np.asarray(nodes), energy


def matching_system(
    parameters: np.ndarray,
    quaternions: np.ndarray,
    durations: np.ndarray,
    steps: int,
    *,
    jacobian: bool = False,
) -> tuple[np.ndarray, csc_matrix | None]:
    """Assemble orientation, velocity, acceleration and natural boundary residuals."""
    count = len(parameters)
    ends, sensitivity, _, _ = integrate(parameters, quaternions[:-1], steps, jacobian=jacobian)
    conjugates = quaternions[1:].copy()
    conjugates[:, 1:] *= -1.0
    rotations = left_matrix(conjugates)
    errors = np.einsum("mij,mj->mi", rotations, ends[:, :4])
    orientation, log_derivative = rotation_residual(errors)
    # Harmonic-mean matching weights, without multiplying physical durations
    # (which can underflow/overflow even when both spacings are finite).
    scale = np.maximum(durations[:-1], durations[1:])
    left_duration, right_duration = durations[:-1] / scale, durations[1:] / scale
    total = left_duration + right_duration
    left_scale, right_scale = 2.0 * right_duration / total, 2.0 * left_duration / total
    velocity = left_scale[:, None] * ends[:-1, 4:7] - right_scale[:, None] * parameters[1:, :3]
    acceleration = left_scale[:, None] ** 2 * ends[:-1, 7:10] - right_scale[:, None] ** 2 * parameters[1:, 3:6]
    residual = np.concatenate((
        orientation.ravel(),
        velocity.ravel(),
        acceleration.ravel(),
        parameters[0, 3:6],
        ends[-1, 7:10],
    ))
    if not np.all(np.isfinite(residual)):
        raise RuntimeError("Multiple shooting matching residual is not finite")
    if sensitivity is None:
        return residual, None

    matrix = lil_matrix((9 * count, 9 * count))
    orientation_derivative = log_derivative @ rotations @ sensitivity[:, :4]
    for index in range(count):
        columns = slice(9 * index, 9 * (index + 1))
        matrix[3 * index : 3 * index + 3, columns] = orientation_derivative[index]
        if index + 1 < count:
            row = 3 * count + 3 * index
            matrix[row : row + 3, columns] = left_scale[index] * sensitivity[index, 4:7]
            matrix[row : row + 3, 9 * (index + 1) : 9 * (index + 1) + 3] = -right_scale[index] * np.eye(3)
            row += 3 * (count - 1)
            matrix[row : row + 3, columns] = left_scale[index] ** 2 * sensitivity[index, 7:10]
            matrix[row : row + 3, 9 * (index + 1) + 3 : 9 * (index + 1) + 6] = -(right_scale[index] ** 2) * np.eye(3)
    matrix[-6:-3, 3:6] = np.eye(3)
    matrix[-3:, -9:] = sensitivity[-1, 7:10]
    return residual, matrix.tocsc()


def newton_solve(  # noqa: PLR0913
    parameters: np.ndarray,
    quaternions: np.ndarray,
    durations: np.ndarray,
    steps: int,
    tolerance: float,
    iterations: int,
) -> tuple[np.ndarray, int]:
    """Solve the sparse square matching system with damped Newton steps."""
    for iteration in range(iterations + 1):
        residual, jacobian = matching_system(parameters, quaternions, durations, steps, jacobian=True)
        if np.max(np.abs(residual)) <= tolerance:
            return parameters, iteration
        if iteration == iterations:
            break
        direction = spsolve(jacobian, -residual).reshape(-1, 9)
        if not np.all(np.isfinite(direction)):
            raise RuntimeError("Multiple shooting Jacobian is singular")
        step = 1.0
        for _ in range(20):
            candidate = parameters + step * direction
            try:
                trial, _ = matching_system(candidate, quaternions, durations, steps)
            except RuntimeError:
                step *= 0.5
                continue
            if np.linalg.norm(trial) < (1.0 - 1e-4 * step) * np.linalg.norm(residual):
                parameters = candidate
                break
            step *= 0.5
        else:
            raise RuntimeError("Multiple shooting line search failed to reduce the matching residual")
    raise RuntimeError("Multiple shooting did not converge within max_iterations")

File: cpp/src/shooting_solver.cpp
#include "shooting_solver.hpp"

#include <Eigen/Geometry>
#include <Eigen/SparseLU>
#include <algorithm>
#include <cmath>
#include <stdexcept>

namespace interpolatecpp::quat::shooting_detail {
namespace {

using Sensitivity = Eigen::Matrix<double, 10, 9>;
using Parameters = Eigen::Matrix<double, 9, 1>;
using SparseMatrix = Eigen::SparseMatrix<double>;

Eigen::Matrix3d skew(const Eigen::Vector3d& v) {
    Eigen::Matrix3d result;
    result << 0.0, -v.z(), v.y(), v.z(), 0.0, -v.x(), -v.y(), v.x(), 0.0;
    return result;
}

Eigen::Vector4d values(const Quaternion& q) { return {q.w(), q.x(), q.y(), q.z()}; }

Eigen::Matrix4d left_matrix(const Eigen::Vector4d& q) {
    Eigen::Matrix4d result;
    result(0, 0) = q[0];
    result.block<1, 3>(0, 1) = -q.tail<3>().transpose();
    result.block<3, 1>(1, 0) = q.tail<3>();
    result.block<3, 3>(1, 1) = q[0] * Eigen::Matrix3d::Identity() + skew(q.tail<3>());
    return result;
}

std::pair<Eigen::Vector3d, Eigen::Matrix<double, 3, 4>>
rotation_residual(Eigen::Vector4d q) {
    const double sign = q[0] < 0.0 ? -1.0 : 1.0;
    q *= sign;
    const double scalar = q[0];
    const Eigen::Vector3d vector = q.tail<3>();
    const double radius = vector.norm();
    const double norm_squared = q.squaredNorm();
    double factor;
    double coefficient;
    if (radius < 1e-7) {
        factor = 1.0 / scalar - radius * radius / (3.0 * scalar * scalar * scalar);
        coefficient = -2.0 / (3.0 * scalar * scalar * scalar);
    } else {
        factor = std::atan2(radius, scalar) / radius;
        coefficient = (scalar / norm_squared - factor) / (radius * radius);
    }
    Eigen::Matrix<double, 3, 4> derivative;
    derivative.col(0) = -vector / norm_squared;
    derivative.rightCols<3>() = factor * Eigen::Matrix3d::Identity() + coefficient * vector * vector.transpose();
    return {factor * vector, sign * derivative};
}

State rhs(const State& state, const Eigen::Vector3d& constant,
          const Sensitivity* sensitivity, Sensitivity* derivative) {
    const Eigen::Vector4d q = state.head<4>();
    const Eigen::Vector3d v = state.segment<3>(4);
    const Eigen::Vector3d a = state.tail<3>();
    State result;
    result[0] = -q.tail<3>().dot(v);
    result.segment<3>(1) = q[0] * v + q.tail<3>().cross(v);
    result.segment<3>(4) = a;
    result.tail<3>() = constant - 2.0 * v.cross(a);
    if (sensitivity != nullptr) {
        Eigen::Matrix4d right = Eigen::Matrix4d::Zero();
        right.block<1, 3>(0, 1) = -v.transpose();
        right.block<3, 1>(1, 0) = v;
        right.block<3, 3>(1, 1) = -skew(v);
        derivative->topRows<4>() = right * sensitivity->topRows<4>() +
            left_matrix(q).rightCols<3>() * sensitivity->middleRows<3>(4);
        derivative->middleRows<3>(4) = sensitivity->bottomRows<3>();
        derivative->bottomRows<3>() = 2.0 * skew(a) * sensitivity->middleRows<3>(4) -
            2.0 * skew(v) * sensitivity->bottomRows<3>();
        derivative->block<3, 3>(7, 6) += Eigen::Matrix3d::Identity();
    }
    return result;
}

State rk4_step(const State& state, const Eigen::Vector3d& constant, double step,
               Sensitivity* sensitivity, double* energy) {
    Sensitivity d1, d2, d3, d4, temporary;
    const State k1 = rhs(state, constant, sensitivity, &d1);
    const State y2 = state + 0.5 * step * k1;
    if (sensitivity != nullptr) temporary = *sensitivity + 0.5 * step * d1;
    const State k2 = rhs(y2, constant, sensitivity == nullptr ? nullptr : &temporary, &d2);
    const State y3 = state + 0.5 * step * k2;
    if (sensitivity != nullptr) temporary = *sensitivity + 0.5 * step * d2;
    const State k3 = rhs(y3, constant, sensitivity == nullptr ? nullptr : &temporary, &d3);
    const State y4 = state + step * k3;
    if (sensitivity != nullptr) temporary = *sensitivity + step * d3;
    const State k4 = rhs(y4, constant, sensitivity == nullptr ? nullptr : &temporary, &d4);
    State result = state + (step / 6.0) * (k1 + 2.0 * k2 + 2.0 * k3 + k4);
    const double norm = result.head<4>().norm();
    result.head<4>() /= norm;
    if (sensitivity != nullptr) {
        *sensitivity += (step / 6.0) * (d1 + 2.0 * d2 + 2.0 * d3 + d4);
        const Eigen::Matrix<double, 1, 9> radial = result.head<4>().transpose() * sensitivity->topRows<4>();
        sensitivity->topRows<4>() =
            (sensitivity->topRows<4>() - result.head<4>() * radial) / norm;
    }
    if (energy != nullptr) {
        *energy += (step / 6.0) * (state.tail<3>().squaredNorm() +
            2.0 * y2.tail<3>().squaredNorm() + 2.0 * y3.tail<3>().squaredNorm() + y4.tail<3>().squaredNorm());
    }
    return result;
}

struct Integration {
    State state;
    Sensitivity sensitivity;
    std::vector<State> nodes;
    double energy = 0.0;
};

Integration integrate(const Parameters& parameters, const Quaternion& start,
                       int steps, bool jacobian, bool store_nodes = false) {
    Integration result;
    result.state << values(start), parameters.head<6>();
    result.sensitivity.setZero();
    result.sensitivity.block<6, 6>(4, 0).setIdentity();
    if (store_nodes) {
        result.nodes.reserve(static_cast<std::size_t>(steps) + 1);
        result.nodes.push_back(result.state);
    }
    for (int index = 0; index < steps; ++index) {
        result.state = rk4_step(result.state, parameters.tail<3>(), 1.0 / static_cast<double>(steps),
                                jacobian ? &result.sensitivity : nullptr, store_nodes ? &result.energy : nullptr);
        if (!result.state.allFinite()) throw std::runtime_error("Shooting integration diverged");
        if (store_nodes) result.nodes.push_back(result.state);
    }
    return result;
}

struct System {
    Eigen::VectorXd residual;
    SparseMatrix jacobian;
};

System matching_system(const Eigen::MatrixXd& parameters, const std::vector<Quaternion>& quaternions,
                        const std::vector<double>& durations, int steps, bool jacobian) {
    const Eigen::Index count = parameters.rows();
    System result{Eigen::VectorXd::Zero(9 * count), SparseMatrix(9 * count, 9 * count)};
    std::vector<Eigen::Triplet<double>> triplets;
    const auto add_block = [&triplets](Eigen::Index row, Eigen::Index column, const auto& block) {
        for (Eigen::Index i = 0; i < block.rows(); ++i) {
            for (Eigen::Index j = 0; j < block.cols(); ++j) {
                triplets.emplace_back(static_cast<int>(row + i), static_cast<int>(column + j), block(i, j));
            }
        }
    };
    for (Eigen::Index index = 0; index < count; ++index) {
        const std::size_t segment = static_cast<std::size_t>(index);
        const auto end = integrate(parameters.row(index).transpose(), quaternions[segment], steps, jacobian);
        const Eigen::Matrix4d rotation = left_matrix(values(quaternions[segment + 1].conjugate()));
        const auto [orientation, log_derivative] = rotation_residual(rotation * end.state.head<4>());
        result.residual.segment<3>(3 * index) = orientation;
        if (jacobian) {
            const Eigen::Matrix<double, 3, 9> block = log_derivative * rotation * end.sensitivity.topRows<4>();
            add_block(3 * index, 9 * index, block);
        }
        if (index + 1 < count) {
            // Scale before forming harmonic-mean weights: physical duration
            // products can overflow or underflow for otherwise valid inputs.
            const double scale = std::max(durations[segment], durations[segment + 1]);
            const double h_left = durations[segment] / scale;
            const double h_right = durations[segment + 1] / scale;
            const double left = 2.0 * h_right / (h_left + h_right);
            const double right = 2.0 * h_left / (h_left + h_right);
            Eigen::Index row = 3 * count + 3 * index;
            result.residual.segment<3>(row) = left * end.state.segment<3>(4) -
                right * parameters.block<1, 3>(index + 1, 0).transpose();
            if (jacobian) {
                add_block(row, 9 * index, (left * end.sensitivity.middleRows<3>(4)).eval());
                add_block(row, 9 * (index + 1), (-right * Eigen::Matrix3d::Identity()).eval());
            }
            row += 3 * (count - 1);
            result.residual.segment<3>(row) = left * left * end.state.tail<3>() -
                right * right * parameters.block<1, 3>(index + 1, 3).transpose();
            if (jacobian) {
                add_block(row, 9 * index, (left * left * end.sensitivity.bottomRows<3>()).eval());
                add_block(row, 9 * (index + 1) + 3, (-right * right * Eigen::Matrix3d::Identity()).eval());
            }
        } else {
            result.residual.tail<3>() = end.state.tail<3>();
            if (jacobian) add_block(9 * count - 3, 9 * index, end.sensitivity.bottomRows<3>());
        }
    }
    result.residual.segment<3>(9 * count - 6) = parameters.block<1, 3>(0, 3).transpose();
    if (!result.residual.allFinite()) {
        throw std::runtime_error("Multiple shooting matching residual is not finite");
    }
    if (jacobian) {
        add_block(9 * count - 6, 3, Eigen::Matrix3d::Identity());
        result.jacobian.setFromTriplets(triplets.begin(), triplets.end());
    }
    return result;
}

int newton_solve(Eigen::MatrixXd& parameters, const std::vector<Quaternion>& quaternions,
                 const std::vector<double>& durations, int steps, double tolerance, int iterations) {
    Eigen::SparseLU<SparseMatrix> solver;
    for (int iteration = 0; iteration <= iterations; ++iteration) {
        const System current = matching_system(parameters, quaternions, durations, steps, true);
        if (current.residual.cwiseAbs().maxCoeff() <= tolerance) return iteration;
        if (iteration == iterations) break;
        solver.compute(current.jacobian);
        if (solver.info() != Eigen::Success) throw std::runtime_error("Multiple shooting Jacobian is singular");
        const Eigen::VectorXd solution = solver.solve(-current.residual);
        if (solver.info() != Eigen::Success || !solution.allFinite()) {
            throw std::runtime_error("Multiple shooting Jacobian solve failed");
        }
        Eigen::MatrixXd direction(parameters.rows(), 9);
        for (Eigen::Index index = 0; index < parameters.rows(); ++index) {
            direction.row(index) = solution.segment<9>(9 * index).transpose();
        }
        double step = 1.0;
        bool accepted = false;
        for (int backtrack = 0; backtrack < 20; ++backtrack) {
            Eigen::MatrixXd candidate = parameters + step * direction;
            try {
                const auto trial = matching_system(candidate, quaternions, durations, steps, false);
                if (trial.residual.norm() < (1.0 - 1e-4 * step) * current.residual.norm()) {
                    parameters = std::move(candidate);
                    accepted = true;
                    break;
                }
            } catch (const std::runtime_error&) {
                // A divergent trial integration is rejected by backtracking.
            }
            step *= 0.5;
        }
        if (!accepted) throw std::runtime_error("Multiple shooting line search failed to reduce the matching residual");
    }
    throw std::runtime_error("Multiple shooting did not converge within max_iterations");
}

}  // namespace

State evaluate_step(const State& state, const Eigen::Vector3d& constant, double step) {
    return rk4_step(state, constant, step, nullptr, nullptr);
}

Solution solve(const std::vector<Quaternion>& quaternions,
               const std::vector<double>& durations, const ShootingConfig& config) {
    const Eigen::Index count = static_cast<Eigen::Index>(durations.size());
    Solution result;
    result.parameters = Eigen::MatrixXd::Zero(count, 9);
    for (Eigen::Index index = 0; index < count; ++index) {
        const auto segment = static_cast<std::size_t>(index);
        const Eigen::Vector4d relative = left_matrix(values(quaternions[segment].conjugate())) * values(quaternions[segment + 1]);
        result.parameters.block<1, 3>(index, 0) = rotation_residual(relative).first.transpose();
    }
    int steps = config.integration_steps;
    while (steps < config.max_integration_steps) {
        result.iterations += newton_solve(result.parameters, quaternions, durations, steps,
                                          config.tolerance * 0.1, config.max_iterations - result.iterations);
        const int finer = steps > config.max_integration_steps / 2 ? config.max_integration_steps : 2 * steps;
        const auto verification = matching_system(result.parameters, quaternions, durations, finer, false);
        result.residual = verification.residual.cwiseAbs().maxCoeff();
        if (result.residual <= config.tolerance) {
            result.steps = finer;
            break;
        }
        steps = finer;
    }
    if (result.steps == 0) {
        throw std::runtime_error("Multiple shooting integration accuracy exceeds tolerance at max_integration_steps");
    }
    result.nodes.reserve(durations.size());
    for (Eigen::Index index = 0; index < count; ++index) {
        const auto segment = static_cast<std::size_t>(index);
        auto curve = integrate(result.parameters.row(index).transpose(), quaternions[segment], result.steps, false, true);
        result.nodes.push_back(std::move(curve.nodes));
        const double duration = durations[segment];
        result.energy += 4.0 * curve.energy / duration / duration / duration;
    }
    return result;
}

}  // namespace interpolatecpp::quat::shooting_detail
